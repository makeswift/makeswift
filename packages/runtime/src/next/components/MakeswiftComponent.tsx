'use client'

import { memo, Suspense, useMemo } from 'react'
import { componentDocumentToRootEmbeddedDocument, MakeswiftComponentSnapshot } from '../client'
import { DocumentRoot } from '../../runtimes/react/components/DocumentRoot'
import { useCacheData } from '../../runtimes/react/hooks/use-cache-data'
import { useRegisterDocument } from '../../runtimes/react/hooks/use-register-document'

type Props = {
  snapshot: MakeswiftComponentSnapshot
  label: string
  type: string
}

export const MakeswiftComponent = memo(({ snapshot, label, type }: Props) => {
  useCacheData(snapshot.cacheData)

  const rootDocument = useMemo(
    () =>
      componentDocumentToRootEmbeddedDocument({
        document: snapshot.document,
        name: label,
        type,
      }),
    [snapshot, label, type],
  )

  useRegisterDocument(rootDocument)

  return (
    <Suspense>
      <DocumentRoot rootDocument={rootDocument} />
    </Suspense>
  )
})
