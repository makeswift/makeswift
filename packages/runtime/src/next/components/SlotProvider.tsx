import { createContext, ReactNode } from 'react'

export const PropsContext = createContext<{
  fallback: ReactNode
}>({
  fallback: null,
})

function SlotProvider({
  value,
  children,
}: {
  value: { fallback: ReactNode }
  children: ReactNode
}) {
  return <PropsContext.Provider value={value}>{children}</PropsContext.Provider>
}

export default SlotProvider
