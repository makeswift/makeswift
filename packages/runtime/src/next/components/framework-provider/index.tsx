'use client'

import { type PropsWithChildren, useMemo } from 'react'
import NextImage from 'next/image'

import { useIsPagesRouter } from '../../hooks/use-is-pages-router'
import {
  FrameworkContext,
  versionedFetch,
} from '../../../runtimes/react/components/framework-context'

import { type SiteVersion } from '../../../api/site-version'
import { MAKESWIFT_CACHE_TAG } from '../../cache'

import { context as appRouterContext } from './app-router'
import { context as pagesRouterContext } from './pages-router'
import { Link } from './link'

export function FrameworkProvider({
  children,
  forcePagesRouter,
}: PropsWithChildren<{ forcePagesRouter?: boolean }>) {
  const isPagesRouter = useIsPagesRouter() || forcePagesRouter
  const context = useMemo<FrameworkContext>(
    () => ({
      ...(isPagesRouter ? pagesRouterContext : appRouterContext),
      Image: NextImage,
      Link,
      versionedFetch: (siteVersion: SiteVersion | null) => (url, init) =>
        versionedFetch(siteVersion)(url, { ...init, next: { tags: [MAKESWIFT_CACHE_TAG] } }),
    }),
    [isPagesRouter],
  )

  return <FrameworkContext.Provider value={context}>{children}</FrameworkContext.Provider>
}
