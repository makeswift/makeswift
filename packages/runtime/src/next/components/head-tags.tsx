import { ComponentPropsWithoutRef } from 'react'
// import { useIsPagesRouter } from '../hooks/use-is-pages-router'
// import Head from 'next/head'

type PageTitleProps = Omit<ComponentPropsWithoutRef<'title'>, 'children'> & {
  children?: string
}

export function PageTitle({ children, ...props }: PageTitleProps) {
  // DECOUPLE_TODO:
  // if (useIsPagesRouter()) {
  //   return (
  //     <Head>
  //       <title {...props}>{children}</title>
  //     </Head>
  //   )
  // }
  return <title {...props}>{children}</title>
}

export function PageMeta(props: ComponentPropsWithoutRef<'meta'>) {
  // if (useIsPagesRouter()) {
  //   return (
  //     <Head>
  //       <meta {...props} />
  //     </Head>
  //   )
  // }
  return <meta {...props} />
}

type ReactCanaryLinkProps = {
  precedence?: 'reset' | 'low' | 'medium' | 'high'
}
type PageLinkProps = ComponentPropsWithoutRef<'link'> & ReactCanaryLinkProps

export function PageLink({ precedence, ...props }: PageLinkProps) {
  // if (useIsPagesRouter()) {
  //   return (
  //     <Head>
  //       <link {...props} />
  //     </Head>
  //   )
  // }
  // @ts-expect-error: currently, react types don't include a prop definition
  // for `precedence` since it's a React Canary feature. See
  // https://react.dev/reference/react-dom/components/link#props. The
  // functionality is supported because Next.js App Router uses React Canary.
  return <link {...props} precedence={precedence} />
}

type ReactCanaryStyleProps = {
  precedence?: 'reset' | 'low' | 'medium' | 'high'
  href?: string
}
type PageStyleProps = ComponentPropsWithoutRef<'style'> & ReactCanaryStyleProps

export function PageStyle({ children, precedence, href, ...props }: PageStyleProps) {
  // if (useIsPagesRouter()) {
  //   return (
  //     <Head>
  //       <style {...props}>{children}</style>
  //     </Head>
  //   )
  // }
  return (
    // @ts-expect-error: currently react types don't include prop definitions
    // for `precedence` & `href` since it's a React Canary feature. See
    // https://react.dev/reference/react-dom/components/style#props. The
    // functionality is supported because Next.js App Router uses React Canary.
    <style {...props} precedence={precedence} href={href}>
      {children}
    </style>
  )
}
