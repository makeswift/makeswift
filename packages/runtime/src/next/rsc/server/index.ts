import 'server-only'

export { NextRSCMakeswiftPage as ExperimentalMakeswiftPage } from './makeswift-page'
export { NextRSCServerProvider as ExperimentalServerProvider } from './server-provider'
