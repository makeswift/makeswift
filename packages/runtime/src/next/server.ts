export { MakeswiftApiHandler } from './api-handler'
export type { Manifest, Font } from '../api-handler'
export type { MakeswiftApiHandlerResponse } from './api-handler'
export { getSiteVersion } from './draft-mode'
