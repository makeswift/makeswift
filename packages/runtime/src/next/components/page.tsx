'use client'

import { Suspense, memo, useMemo } from 'react'

import { useDispatch } from '../../runtimes/react/hooks/use-dispatch'
import { useUniversalDispatch } from '../../runtimes/react/hooks/use-universal-dispatch'
import { useCacheData } from '../../runtimes/react/hooks/use-cache-data'

import { Page as PageComponent } from '../../components/page'
import {
  type MakeswiftPageSnapshot,
  type MakeswiftPageDocument,
  pageToRootDocument,
} from '../client'
import * as ReactPage from '../../state/react-page'
import { registerDocumentsEffect } from '../../state/actions'

export type PageProps = {
  snapshot: MakeswiftPageSnapshot
}

function useRegisterPageDocument(pageDocument: MakeswiftPageDocument): ReactPage.Document {
  const dispatch = useDispatch()
  const rootDocuments: [ReactPage.Document] = useMemo(
    () => [pageToRootDocument(pageDocument)],
    [pageDocument],
  )

  useUniversalDispatch(dispatch, registerDocumentsEffect, [rootDocuments])

  return rootDocuments[0]
}

export const Page = memo(({ snapshot, ...props }: PageProps) => {
  if ('runtime' in props) {
    throw new Error(
      `The \`runtime\` prop is no longer supported in the \`@makeswift/runtime\` \`Page\` component as of \`0.15.0\`.
See our docs for more information on what's changed and instructions to migrate: https://docs.makeswift.com/migrations/0.15.0`,
    )
  }

  useCacheData(snapshot.cacheData)

  const rootDocument = useRegisterPageDocument(snapshot.document)

  return (
    <Suspense>
      {/* We use a key here to reset the Snippets state in the PageMeta component */}
      <PageComponent
        key={snapshot.document.data.key}
        page={snapshot.document}
        rootDocument={rootDocument}
      />
    </Suspense>
  )
})
