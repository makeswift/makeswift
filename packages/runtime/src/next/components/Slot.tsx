'use client'

import { ReactNode, useMemo } from 'react'
import { MakeswiftComponentSnapshot } from '../client'
import { MakeswiftComponentType } from '../../components'
import { MakeswiftComponent } from './MakeswiftComponent'
import SlotProvider from './SlotProvider'

type Props = {
  label: string
  snapshot: MakeswiftComponentSnapshot
  fallback?: ReactNode
}

export const Slot = ({ label, snapshot, fallback }: Props) => {
  const contextValue = useMemo(() => ({ fallback }), [fallback])

  return (
    <SlotProvider value={contextValue}>
      <MakeswiftComponent snapshot={snapshot} label={label} type={MakeswiftComponentType.Slot} />
    </SlotProvider>
  )
}

export default Slot
