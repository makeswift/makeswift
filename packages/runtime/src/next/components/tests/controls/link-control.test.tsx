/** @jest-environment jsdom */
import { Link } from '../../../../controls'
import { testPageControlPropRendering } from './page-control-prop-rendering'
import { APIResourceType } from '../../../../api'
import { http, HttpResponse } from 'msw'
import { server } from '../../../../mocks/server'
import { MakeswiftComponentType } from '../../../../components'
import { ElementID } from '@makeswift/prop-controllers'
import { ReactRuntime } from '../../../../react'

function Button({ id }: { id?: string }) {
  return <button id={id} />
}

const pathnameSlicesBaseUrl = `/api/makeswift/page-pathname-slices`

const pageId = '[test-page-id]'

const pagePathnameSlice = {
  __typename: APIResourceType.PagePathnameSlice,
  id: pageId,
  pathname: 'test-pathname',
}

const localizedPagePathnameSlice = {
  __typename: APIResourceType.PagePathnameSlice,
  id: pageId,
  basePageId: '[test-base-page-id]',
  pathname: 'fr-FR/test-localized-pathname',
}

const openPageLink = {
  type: 'OPEN_PAGE',
  payload: { pageId, openInNewTab: false },
} as const

const openUrlLink = {
  type: 'OPEN_URL',
  payload: { url: 'https://example.com', openInNewTab: false },
} as const

const sendEmailLink = {
  type: 'SEND_EMAIL',
  payload: { to: 'bob@gmail.com', subject: 'hello', body: 'world' },
} as const

const callPhoneLink = {
  type: 'CALL_PHONE',
  payload: { phoneNumber: '555-555-5555' },
} as const

const scrollToElementLink = {
  type: 'SCROLL_TO_ELEMENT',
  payload: {
    elementIdConfig: { elementKey: 'key', propName: 'prop' },
    block: 'start',
  },
} as const

describe('Page', () => {
  test.each([openUrlLink, sendEmailLink, callPhoneLink, scrollToElementLink, null] as const)(
    'Link control data %s',
    async value => {
      await testPageControlPropRendering(Link(), { value })
    },
  )

  describe.each([null, 'fr-FR'])('Link control data OPEN_PAGE for locale %s', locale => {
    const pathnameSlice = locale ? localizedPagePathnameSlice : pagePathnameSlice

    test('resolves default link attributes when unable to find pathname slice', async () => {
      await testPageControlPropRendering(Link(), {
        value: openPageLink,
        locale,
        cacheData: {
          PagePathnameSlice: [{ id: pageId, value: null }],
        },
      })
    })

    test('fetches pathname slice when not in cache', async () => {
      server.use(
        http.get(`${pathnameSlicesBaseUrl}/${openPageLink.payload.pageId}`, ({ request }) => {
          const requestedLocale = new URL(request.url).searchParams.get('locale')
          return requestedLocale === locale
            ? HttpResponse.json(pathnameSlice)
            : HttpResponse.json(null)
        }),
      )
      await testPageControlPropRendering(Link(), {
        value: openPageLink,
        locale,
        expectedRenders: 2,
      })
    })

    test('renders page link when page pathname slice is in cache', async () => {
      await testPageControlPropRendering(Link(), {
        value: openPageLink,
        locale,
        cacheData: {
          PagePathnameSlice: [{ id: pageId, value: pathnameSlice }],
        },
        expectedRenders: 1,
      })
    })
  })

  describe('Link control data SCROLL_TO_ELEMENT', () => {
    test('resolves default link attributes when elementIdConfig is not provided', async () => {
      await testPageControlPropRendering(Link(), {
        value: scrollToElementLink,
      })
    })

    test('resolves scroll target when element key is on screen', async () => {
      await testPageControlPropRendering(Link(), {
        value: scrollToElementLink,
        expectedRenders: 1,
        registerComponents: (runtime: ReactRuntime) => {
          runtime.registerComponent(Button, {
            type: MakeswiftComponentType.Button,
            label: 'Button',
            props: {
              id: ElementID(),
            },
          })
        },
        rootElements: [
          {
            key: 'key',
            type: MakeswiftComponentType.Button,
            props: {
              id: 'test-element-id',
            },
          },
        ],
      })
    })
  })
})
