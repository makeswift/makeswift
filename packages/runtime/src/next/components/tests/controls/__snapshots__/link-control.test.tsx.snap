// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Link control data { type: 'CALL_PHONE', payload: [Object] }: resolvedValue 1`] = `
{
  "href": "tel:555-555-5555",
}
`;

exports[`Page Link control data { type: 'CALL_PHONE', payload: [Object] }: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {},
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "payload": {
                    "phoneNumber": "555-555-5555",
                  },
                  "type": "CALL_PHONE",
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`Page Link control data { type: 'OPEN_URL', payload: [Object] }: resolvedValue 1`] = `
{
  "href": "https://example.com",
  "target": "_self",
}
`;

exports[`Page Link control data { type: 'OPEN_URL', payload: [Object] }: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {},
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "payload": {
                    "openInNewTab": false,
                    "url": "https://example.com",
                  },
                  "type": "OPEN_URL",
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`Page Link control data { type: 'SCROLL_TO_ELEMENT', payload: [Object] }: resolvedValue 1`] = `
{
  "href": "#",
}
`;

exports[`Page Link control data { type: 'SCROLL_TO_ELEMENT', payload: [Object] }: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {},
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "payload": {
                    "block": "start",
                    "elementIdConfig": {
                      "elementKey": "key",
                      "propName": "prop",
                    },
                  },
                  "type": "SCROLL_TO_ELEMENT",
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`Page Link control data { type: 'SEND_EMAIL', payload: [Object] }: resolvedValue 1`] = `
{
  "href": "mailto:bob@gmail.com?subject=hello&body=world",
}
`;

exports[`Page Link control data { type: 'SEND_EMAIL', payload: [Object] }: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {},
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "payload": {
                    "body": "world",
                    "subject": "hello",
                    "to": "bob@gmail.com",
                  },
                  "type": "SEND_EMAIL",
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`Page Link control data OPEN_PAGE for locale fr-FR fetches pathname slice when not in cache: resolvedValue 1`] = `
{
  "href": "/fr-FR/test-localized-pathname",
  "target": "_self",
}
`;

exports[`Page Link control data OPEN_PAGE for locale fr-FR fetches pathname slice when not in cache: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {},
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "payload": {
                    "openInNewTab": false,
                    "pageId": "[test-page-id]",
                  },
                  "type": "OPEN_PAGE",
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "id": "test-page-id",
    "locale": "fr-FR",
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`Page Link control data OPEN_PAGE for locale fr-FR renders page link when page pathname slice is in cache: resolvedValue 1`] = `
{
  "href": "/fr-FR/test-localized-pathname",
  "target": "_self",
}
`;

exports[`Page Link control data OPEN_PAGE for locale fr-FR renders page link when page pathname slice is in cache: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {
      "PagePathnameSlice": [
        {
          "id": "[test-page-id]",
          "locale": "fr-FR",
          "value": {
            "__typename": "PagePathnameSlice",
            "basePageId": "[test-base-page-id]",
            "id": "[test-page-id]",
            "pathname": "fr-FR/test-localized-pathname",
          },
        },
      ],
    },
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "payload": {
                    "openInNewTab": false,
                    "pageId": "[test-page-id]",
                  },
                  "type": "OPEN_PAGE",
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "id": "test-page-id",
    "locale": "fr-FR",
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`Page Link control data OPEN_PAGE for locale fr-FR resolves default link attributes when unable to find pathname slice: resolvedValue 1`] = `
{
  "href": "#",
  "target": "_self",
}
`;

exports[`Page Link control data OPEN_PAGE for locale fr-FR resolves default link attributes when unable to find pathname slice: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {
      "PagePathnameSlice": [
        {
          "id": "[test-page-id]",
          "locale": "fr-FR",
          "value": null,
        },
      ],
    },
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "payload": {
                    "openInNewTab": false,
                    "pageId": "[test-page-id]",
                  },
                  "type": "OPEN_PAGE",
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "id": "test-page-id",
    "locale": "fr-FR",
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`Page Link control data OPEN_PAGE for locale null fetches pathname slice when not in cache: resolvedValue 1`] = `
{
  "href": "/test-pathname",
  "target": "_self",
}
`;

exports[`Page Link control data OPEN_PAGE for locale null fetches pathname slice when not in cache: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {},
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "payload": {
                    "openInNewTab": false,
                    "pageId": "[test-page-id]",
                  },
                  "type": "OPEN_PAGE",
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`Page Link control data OPEN_PAGE for locale null renders page link when page pathname slice is in cache: resolvedValue 1`] = `
{
  "href": "/test-pathname",
  "target": "_self",
}
`;

exports[`Page Link control data OPEN_PAGE for locale null renders page link when page pathname slice is in cache: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {
      "PagePathnameSlice": [
        {
          "id": "[test-page-id]",
          "locale": null,
          "value": {
            "__typename": "PagePathnameSlice",
            "id": "[test-page-id]",
            "pathname": "test-pathname",
          },
        },
      ],
    },
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "payload": {
                    "openInNewTab": false,
                    "pageId": "[test-page-id]",
                  },
                  "type": "OPEN_PAGE",
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`Page Link control data OPEN_PAGE for locale null resolves default link attributes when unable to find pathname slice: resolvedValue 1`] = `
{
  "href": "#",
  "target": "_self",
}
`;

exports[`Page Link control data OPEN_PAGE for locale null resolves default link attributes when unable to find pathname slice: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {
      "PagePathnameSlice": [
        {
          "id": "[test-page-id]",
          "locale": null,
          "value": null,
        },
      ],
    },
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "payload": {
                    "openInNewTab": false,
                    "pageId": "[test-page-id]",
                  },
                  "type": "OPEN_PAGE",
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`Page Link control data SCROLL_TO_ELEMENT resolves default link attributes when elementIdConfig is not provided: resolvedValue 1`] = `
{
  "href": "#",
}
`;

exports[`Page Link control data SCROLL_TO_ELEMENT resolves default link attributes when elementIdConfig is not provided: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {},
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "payload": {
                    "block": "start",
                    "elementIdConfig": {
                      "elementKey": "key",
                      "propName": "prop",
                    },
                  },
                  "type": "SCROLL_TO_ELEMENT",
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`Page Link control data SCROLL_TO_ELEMENT resolves scroll target when element key is on screen: resolvedValue 1`] = `
{
  "href": "#test-element-id",
}
`;

exports[`Page Link control data SCROLL_TO_ELEMENT resolves scroll target when element key is on screen: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {},
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "payload": {
                    "block": "start",
                    "elementIdConfig": {
                      "elementKey": "key",
                      "propName": "prop",
                    },
                  },
                  "type": "SCROLL_TO_ELEMENT",
                },
              },
              "type": "TestComponent",
            },
            {
              "key": "key",
              "props": {
                "id": "test-element-id",
              },
              "type": "./components/Button/index.js",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`Page Link control data null: resolvedValue 1`] = `
{
  "href": "#",
}
`;

exports[`Page Link control data null: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {},
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": null,
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;
